packages/hub/services/prisma-manager.ts

```diff
@@ -0,0 +1,97 @@
+import { Prisma, PrismaClient } from '@prisma/client';
+import config from 'config';
+import { PrismaTestingHelper } from '@chax-at/transactional-prisma-testing';
+
+type push_notification_registrations_getter = Prisma.push_notification_registrationsDelegate<any>;
+// FIXME this is <GlobalRejectSettings> but how to import?
+
+interface ExtendedPushNotificationRegistrations extends push_notification_registrations_getter {
+  upsertByOwnerAndPushClient({
+    id,
+    owner_address,
+    push_client_id,
+    disabled_at,
+  }: {
+    id: string;
+    owner_address: string;
+    push_client_id: string;
+    disabled_at: Date | null;
+  }): ReturnType<push_notification_registrations_getter['upsert']>;
+}
+
+export interface ExtendedPrismaClient extends PrismaClient {
+  push_notification_registrations: ExtendedPushNotificationRegistrations;
+}
+
+export default class PrismaManager {
+  private client?: ExtendedPrismaClient;
+  private prismaTestingHelper?: PrismaTestingHelper<PrismaClient>;
+
+  dbConfig: Record<string, any> = config.get('db');
+
+  async getClient() {
+    if (!this.client) {
+      let client = new PrismaClient({
+        datasources: { db: { url: this.dbConfig.url } },
+        log: this.dbConfig.prismaLog,
+      });
+
+      if (this.dbConfig.useTransactionalRollbacks) {
+        this.prismaTestingHelper = new PrismaTestingHelper(client);
+        await this.prismaTestingHelper.startNewTransaction();
+        client = this.prismaTestingHelper.getProxyClient();
+      }
+
+      Object.assign(client.push_notification_registrations, {
+        upsertByOwnerAndPushClient({
+          id,
+          owner_address,
+          push_client_id,
+          disabled_at = null,
+        }: {
+          id: string;
+          owner_address: string;
+          push_client_id: string;
+          disabled_at: Date | null;
+        }) {
+          return client.push_notification_registrations.upsert(
+            Prisma.validator<Prisma.push_notification_registrationsUpsertArgs>()({
+              where: {
+                owner_address_push_client_id: {
+                  owner_address,
+                  push_client_id,
+                },
+              },
+              create: {
+                id: id,
+                owner_address,
+                push_client_id,
+                disabled_at,
+              },
+              update: {
+                disabled_at,
+              },
+            })
+          );
+        },
+      });
+
+      this.client = client as ExtendedPrismaClient;
+    }
+
+    return this.client;
+  }
+
+  async teardown() {
+    this.prismaTestingHelper?.rollbackCurrentTransaction();
+    // FIXME I hoped the below would address this warning but no:
+    // warn(prisma-client) There are already 10 instances of Prisma Client actively running.
+    return this.client?.$disconnect();
```

Comment:
This is discussed [here](https://www.prisma.io/docs/guides/database/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices#solution). Should we have some `global` hack for testing? I think it only happens then because the service is constantly being recreated.

Reply:
I really would like to avoid the global thing. However, it seems that the warning is emitted based on module-scoped data that cannot be reset and actually would be the source of a memory leak itself the way we are using PrismaClient: https://github.com/prisma/prisma/blob/67dabe14b33748d64d4c08a16fa1871836e4cedd/packages/engine-core/src/library/LibraryEngine.ts#L96-L109 🙀

Reply:
I’m sure we can find a way to accomplish this without excessive hacks. It does seem like we should be able to construct the client once in tests and go with it forever. Noted in 4254.